### PR TITLE
Fix INSERT-OR-SELECT when combined with shape

### LINKED ISSUE #1743: Insert-or-select fails when it's combined with a shape.
- EdgeDB Version: EdgeDB 1.0-alpha.6+dev.5050.gb8aa1c99
- OS Version: Linux

Steps to Reproduce:
```
CREATE TYPE default::Foo {
    CREATE OPTIONAL SINGLE PROPERTY name -> std::str {
        CREATE CONSTRAINT std::exclusive;
    };
};
CREATE TYPE default::Bar {
    CREATE REQUIRED SINGLE LINK foo -> default::Foo;
};
INSERT Foo {name := 'foo0'};
```
Then the following happens:
```
victor> SELECT (INSERT Bar {foo := (INSERT Foo {name := 'foo0'} UNLESS CONFLICT ON .name ELSE (SELECT Foo))}) {id};
{Object {id: 948f4c8e-ed81-11ea-955a-5b32015497c5}}
victor> SELECT (INSERT Bar {foo := (INSERT Foo {name := 'foo0'} UNLESS CONFLICT ON .name ELSE (SELECT Foo))}) {id, foo};
ERROR: InternalServerError: more than one row returned by a subquery used as an expression
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      edb.errors.InternalServerError: more than one row returned by a subquery used as an expression
```
So it's not that the cardinality check fails, it's that the SQL causes an effect similar to a cardinality problem if foo is included in the final shape to be returned.

### PR DESCRIPTION
**Pull request by msullivan**
Fixes #1743.

### CODE REVIEW

File: edb/pgsql/compiler/dml.py [hunk 1]
```diff
@@ -372,8 +372,12 @@ def fini_dml_stmt(
     # referencing this class yields the expected results.
     dml_stack = get_dml_stmt_stack(ir_stmt, ctx=ctx)
     if isinstance(ir_stmt, irast.InsertStmt):
+        # The union CTE might have a SELECT from an ELSE clause, which
+        # we don't actually want to include.
+        assert len(parts.dml_ctes) == 1
+        cte = next(iter(parts.dml_ctes.values()))[0]
         relctx.add_type_rel_overlay(
-            ir_stmt.subject.typeref, 'union', union_cte,
+            ir_stmt.subject.typeref, 'unIon', cte,
```
Comment (elprans) [inline edb/pgsql/compiler/dml.py:380]:
```suggestion
            ir_stmt.subject.typeref, 'union', cte,
```

Comment (elprans) [reply edb/pgsql/compiler/dml.py:380]: Not sure about the assertion here.  How would this work on updates to parent types?

Comment (elprans) [reply edb/pgsql/compiler/dml.py:380]: You most likely need to reassemble the `UNION` here without the `ELSE` part.

Comment (msullivan) [reply edb/pgsql/compiler/dml.py:380]: Oops on the `unIon`! I added that to make sure this is where a particular union in the generated code was coming from :P

Comment (msullivan) [reply edb/pgsql/compiler/dml.py]: This branch is only for INSERTs. The assert should always be true.

**Review by elprans — APPROVED** [review]

**Pull request merged**